pkg/parser/sqlflow_parser.go

```diff
@@ -123,6 +123,24 @@ func thirdPartyParse(dialect, program string) ([]*SQLFlowStmt, int, error) {
 	return spr, i, nil
 }
 
+func findNextSQLStatement(statements string) (string, error) {
```

Comment:
Can we don't skip over the parser and call the lexer directly?  In my previous PRs, I made the `parseSQLStatement` in extended_syntax.y able to return index as the external parsers. I think we can use the ability to get rid of the necessity to call the lexer directly.

Reply:
I was trying to use the return index from extended_syntax.y but couldn't find a solution.

Consider we parsing `to train with ... label l into some_table; select * from some_other_table`. The extended syntax parser will return an error index at the character `s` of `select`. So we should try to parse the string before the error index (i.e. `to train with ... label l into some_table; `), and this time the parsing is successful. To put this logic into code, we have

```go
pr, idx, err := parseSQLFlowStmt(statements)

if err != nil {
    pr, _, err := parseSQLFlowStmt(statements[:idx])
    if err != nil {
        return nil, err
    }
    return pr, err
}

return pr, err
```

However, this logic fails on cases like `to train a with a = b label c into d select * from e`. Our parser won't be raising error while it should be.

Reply:
One solution I can think of is to make the extended syntax parser incapable of recognizing `;`, meaning parsing `to train a with a = b label c into d;` will return an error at `;` (this behavior is consistent with Hive parser and Calcite parser).

So when parsing `to train a with a = b label c into d; select * from e;`, the extended syntax parser will return an error indexed at the character `;`. So we check if the index is pointed at `;`, if so, we should try to parse the string before the error index (i.e. `to train a with a = b label c into d`), and this time the parsing is successful. To put this logic into code, we have

```
pr, idx, err := parseSQLFlowStmt(statements)

if statements[idx] == ';' {
    pr, _, err := parseSQLFlowStmt(statements[:idx])
    if err != nil {
        return nil, err
    }
    return pr, err
}

return pr, err
``` 

And this logic will also capture incorrect statement like `to train a with a = b label c into d select * from e`.

Reply:
If we expect our parser errors with `to train a with a = b label c into d select * from e` as we want a `;` between `into d` and `select`, we can simply change the rule `end_of_stmt` to must have a `;` from that `;` is optional.

Reply:
@wangkuiyi originally mentioned the following, I am the one writing it done.

"
The design flaw of https://github.com/sql-machine-learning/sqlflow/pull/1626#discussion_r363575099 is it assumes upper-level logic knows the implementation details of the lower-level.

For a parser, there are four levels of logic:
1. syntax - parser
1. lexical - lexer
1. rune - Unicode
1. bytes/characters

While https://github.com/sql-machine-learning/sqlflow/pull/1626#discussion_r363575099 is suggesting code on level 1 access the logic of level 4.
"